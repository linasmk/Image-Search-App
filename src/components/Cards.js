/* ========= App Dependencies ============= */
import React, { useState } from "react";
import PropTypes from "prop-types";
/* ========= Components =============== */
import CardModal from "@comp/CardModal";
import Loader from "@comp/Loader";
/* ========= Code ============= */

export const Cards = ({ images, topLoader, noResultsError }) => {
  const [cardModalId, setCardModalId] = useState(undefined);
  const openCardModal = (id) => setCardModalId(id);
  const closeCardModal = () => setCardModalId(undefined);

  return (
    <div>
      {topLoader ? (
        <Loader display="block" />
      ) : (
        <section className="card-list">
          {noResultsError ? (
            <section className="no-results">
              <h3 className="no-results__warning">
                Your search keyword did not return any results!
              </h3>
            </section>
          ) : (
            images.map((image) => (
              <div className="card" key={image.id}>
                <img
                  role="presentation"
                  className="card--image"
                  onClick={() => openCardModal(image.id)}
                  alt={image.alt_description}
                  src={image.urls.small}
                  width="50%"
                  height="50%"
                />
                <CardModal
                  openCardModal={cardModalId === image.id}
                  closeCardModal={closeCardModal}
                  altDescription={image.alt_description}
                  imageID={image.id}
                  likes={image.likes}
                  urlsFull={image.urls.full}
                  firstName={image.user.first_name}
                  lastName={image.user.last_name}
                  download={image.links.download}
                />
              </div>
            ))
          )}
        </section>
      )}
    </div>
  );
};
Cards.defaultProps = {
  images: undefined,
  topLoader: false,
  noResultsError: false
};
Cards.propTypes = {
  images: PropTypes.arrayOf(PropTypes.object),
  topLoader: PropTypes.bool,
  noResultsError: PropTypes.bool
};
export default Cards;
