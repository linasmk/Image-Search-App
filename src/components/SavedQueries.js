/* ========= App Dependencies ============= */
import React, { useState } from "react";
import PropTypes from "prop-types";
/* =========== Redux ============== */

/* ========= Components =============== */
import { SavedQueryItem } from "@cont/SavedQueryItem";
/* ========= Code ============= */
export const SavedQueries = ({ savedQueries, onClick }) => {
  const [sliceSavedQueries, setSliceSavedQueries] = useState(10);
  const [queriesListOpen, setQueriesListOpen] = useState(false);

  const toggleQueriesList = () => {
    if (queriesListOpen) {
      setSliceSavedQueries(10);
    } else {
      setSliceSavedQueries(savedQueries.length);
    }
    setQueriesListOpen(!queriesListOpen);
  };
  return (
    <section className="saved-queries">
      <div className="saved-queries__wrapper">
        {savedQueries
          .map((savedQuery) => (
            <SavedQueryItem
              key={savedQuery.id}
              id={savedQuery.id}
              name={savedQuery.name}
              onClick={onClick}
            />
          ))
          .slice(0, sliceSavedQueries)}
        <div className="saved-queries__btn-wrapper">
          <button
            className="saved-queries__btn--show"
            type="button"
            style={
              savedQueries.length > 10
                ? { display: "block" }
                : { display: "none" }
            }
            onClick={toggleQueriesList}>
            {queriesListOpen ? "Show less" : "Show more"}
          </button>
        </div>
      </div>
    </section>
  );
};
SavedQueries.defaultProps = {
  onClick: undefined
};
SavedQueries.propTypes = {
  savedQueries: PropTypes.arrayOf(PropTypes.object).isRequired,
  onClick: PropTypes.func
};
export default SavedQueries;
