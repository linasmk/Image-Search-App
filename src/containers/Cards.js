/* ===== Redux ===== */
import { connect } from "react-redux";

/* ====== Components ===== */
import CardsComponent from "@comp/Cards";

/* ========= Code ========= */
// export const Cards = connect((state) => ({
//   images: state.images
// }))(CardsComponent);
