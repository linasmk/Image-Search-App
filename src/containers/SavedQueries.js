/* ===== Redux ===== */
import { connect } from "react-redux";
import latestQueryFirst from "@selectors/latestQueryFirst";

/* ====== Components ===== */
import SavedQueriesComponent from "@comp/SavedQueries";

/* ========= Code ========= */
export const SavedQueries = connect((state) => ({
  savedQueries: latestQueryFirst(state.savedQueries)
}))(SavedQueriesComponent);
