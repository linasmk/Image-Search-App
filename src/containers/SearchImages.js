/* ===== Redux ===== */
import { connect } from "react-redux";
import { addSavedQuery } from "@actions/savedQueries";
/* ====== Components ===== */
import SearchImagesComponent from "@comp/SearchImages";
/* ========= Code ========= */
const mapDispatchToProps = (dispatch) => ({
  addSavedQuery: (name) => dispatch(addSavedQuery({ name })),
});

export const SearchImages = connect(
  () => ({}),
  mapDispatchToProps,
)(SearchImagesComponent);
