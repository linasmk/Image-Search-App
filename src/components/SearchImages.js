/* ========= App Dependencies ============= */
import React, { useEffect, useState, useRef } from "react";
import PropTypes from "prop-types";
import Unsplash, { toJson } from "unsplash-js"; // eslint-disable-line no-unused-vars

/* ========= Components =============== */
import { SavedQueries } from "@cont/SavedQueries";
import Cards from "@comp/Cards";
import Loader from "@comp/Loader";

import { UnsplashAccessKey } from "@root/accessKey";

/* ========= Code ============= */
export const SearchImages = ({ addSavedQuery }) => {
  // ================ Hooks ==============
  const [query, setQuery] = useState("");
  const [images, setImages] = useState([]);
  const [imagesTotal, setImagesTotal] = useState(0);
  const [pageNumber, setPageNumber] = useState(1);
  const [inputError, setInputError] = useState(false);
  const [noResultsError, setNoResultsError] = useState(false);
  const [excessInputError, setExcessInputError] = useState(false);
  const [topLoader, setTopLoader] = useState(false);
  const [bottomLoader, setBottomLoader] = useState(false);
  const focusOnSearch = useRef(null);

  // ============ Functions ===============================
  const searchPhotos = async (e) => {
    e.preventDefault();
    setNoResultsError(false);
    UnsplashAccessKey.search
      .photos(query, 1, 10)
      .then(toJson)
      .then((json) => {
        setTopLoader(false);
        setImagesTotal(json.total);
        if (json.results <= 0) {
          setNoResultsError(true);
        } else {
          setImages(json.results);
          console.log(json.results);
        }
      });
  };
  const updatePhotos = async () => {
    UnsplashAccessKey.search
      .photos(query, pageNumber, 10)
      .then(toJson)
      .then((json) => {
        setBottomLoader(false);
        setImages([...images, ...json.results]);
      });
  };
  const handleSubmit = (e) => {
    if (query.length === 0) {
      e.preventDefault();
      setInputError(true);
    } else if (query.length > 50) {
      e.preventDefault();
      setInputError(false);
      setExcessInputError(true);
    } else {
      setInputError(false);
      setExcessInputError(false);
      setTopLoader(true);
      searchPhotos(e);
      addSavedQuery(query);
    }
  };
  useEffect(() => {
    updatePhotos(pageNumber);
  }, [pageNumber]);
  const loadMore = () => {
    setPageNumber((prevPageNumber) => prevPageNumber + 1);
    setBottomLoader(true);
  };
  function handleChildQuery(newQuery) {
    setQuery(newQuery);
    focusOnSearch.current.focus();
  }
  const clearInputField = () => {
    setQuery("");
    focusOnSearch.current.focus();
  };
  return (
    <article className="search-images">
      <form className="search-form" onSubmit={handleSubmit}>
        <div className="search-form__input--wrapper">
          <input
            type="text"
            name="query"
            placeholder='Use keywords, ex: "phone"'
            value={query}
            onChange={(e) => setQuery(e.target.value)}
            ref={focusOnSearch}
            className="search-form__input"
          />
          <button
            type="button"
            className="search-form__btn--clear"
            onClick={clearInputField}>
            &#88;
          </button>
        </div>

        <button type="submit" className="search-form__btn--submit">
          Search
        </button>
        <span
          className="search-form__warning"
          style={
            inputError || excessInputError
              ? { display: "block" }
              : { display: "none" }
          }>
          {inputError
            ? "The search field should not stay empty!"
            : "" || excessInputError
            ? "50 characters max length exceeded. Use concise keywords. This is not a poetry app!"
            : ""}
        </span>
      </form>
      <SavedQueries onClick={handleChildQuery} />

      <Cards
        topLoader={topLoader}
        noResultsError={noResultsError}
        images={images}
      />
      <Loader bottomLoader={bottomLoader} display="none" />
      <section className="load-more">
        <div className="load-more__inner-wrapper">
          <p
            style={
              images.length === 0 || noResultsError
                ? { display: "none" }
                : { display: "block" }
            }>
            Currently showing <span>{images.length}</span>
            {images.length > 1 ? " results" : " result"} out of {imagesTotal}
          </p>
          <button
            type="button"
            display="none"
            style={
              images.length === 0 ? { display: "none" } : { display: "block" }
            }
            className="load-more__btn"
            onClick={loadMore}
            disabled={images.length === imagesTotal}>
            Load more
          </button>
        </div>
      </section>
    </article>
  );
};

SearchImages.defaultProps = {
  addSavedQuery: undefined
};
SearchImages.propTypes = {
  addSavedQuery: PropTypes.func
};

export default SearchImages;
