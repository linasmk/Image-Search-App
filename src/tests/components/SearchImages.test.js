/* ========= App Dependencies ============= */
import React from "react";
import { shallow } from "enzyme";
/* ========= Components ============= */
import { SavedQueries } from "@cont/SavedQueries";
import { Cards } from "@comp/Cards";

import SearchImages from "@comp/SearchImages";

/* ========= Code ============= */
let wrapper, addSavedQuery;

beforeEach(() => {
  addSavedQuery = jest.fn();
  wrapper = shallow(<SearchImages addSavedQuery={addSavedQuery} />);
});

it("renders SearchImages component correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

it("shows a form HTML element", () => {
  expect(wrapper.find("form").length).toEqual(1);
});

it("shows a SavedQueries component", () => {
  expect(wrapper.find(SavedQueries).length).toEqual(1);
});

it("shows a Cards component", () => {
  expect(wrapper.find(Cards).length).toEqual(1);
});

it("has a input field that users can type in", () => {
  wrapper.find("input").simulate("change", {
    target: { value: "query" }
  });
  wrapper.update();
  expect(wrapper.find("input").prop("value")).toEqual("query");
});

it("when form is submitted, input field retains the value", () => {
  wrapper.find("input").simulate("change", {
    target: { value: "query" }
  });
  wrapper.update();
  wrapper.find("form").simulate("submit");
  wrapper.update();
  expect(wrapper.find("input").prop("value")).toEqual("query");
});

// it("clears input field with 'search-form__btn--clear' button", () => {
//   wrapper.find("input").simulate("change", {
//     target: { value: "query" }
//   });
//   wrapper.update();
//   expect(wrapper.find("input").prop("value")).toEqual("query");
//   wrapper
//     .find("button")
//     .at(0)
//     .simulate("click", {
//       target: { value: "" }
//     });
//   expect(wrapper.find("input").prop("value")).toEqual("");
//   wrapper.find("input").simulate("focus");
// });
