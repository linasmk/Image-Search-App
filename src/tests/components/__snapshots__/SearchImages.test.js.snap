// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders SearchImages component correctly 1`] = `
<article
  className="search-images"
>
  <form
    className="search-form"
    onSubmit={[Function]}
  >
    <div
      className="search-form__input--wrapper"
    >
      <input
        className="search-form__input"
        name="query"
        onChange={[Function]}
        placeholder="Use keywords, ex: \\"phone\\""
        type="text"
        value=""
      />
      <button
        className="search-form__btn--clear"
        onClick={[Function]}
        type="button"
      >
        X
      </button>
    </div>
    <button
      className="search-form__btn--submit"
      type="submit"
    >
      Search
    </button>
    <span
      className="search-form__warning"
      style={
        Object {
          "display": "none",
        }
      }
    />
  </form>
  <Connect(SavedQueries)
    onClick={[Function]}
  />
  <Cards
    images={Array []}
    noResultsError={false}
    topLoader={false}
  />
  <Loader
    bottomLoader={false}
    display="none"
  />
  <section
    className="load-more"
  >
    <div
      className="load-more__inner-wrapper"
    >
      <p
        style={
          Object {
            "display": "none",
          }
        }
      >
        Currently showing 
        <span>
          0
        </span>
         result
         out of 
        0
      </p>
      <button
        className="load-more__btn"
        disabled={true}
        display="none"
        onClick={[Function]}
        style={
          Object {
            "display": "none",
          }
        }
        type="button"
      >
        Load more
      </button>
    </div>
  </section>
</article>
`;
